File: python/examples/basic_usage.py
"""
Basic usage example for the MyTotalConnectComfort API client.
"""

from mytotalconnectcomfort import Client
from mytotalconnectcomfort.exceptions import (
    AuthenticationError,
    APIError,
    ZoneNotFoundError,
)


def main():
    """Demonstrate basic usage of the API client."""
    
    # Create a client instance
    client = Client()
    
    # Login with your credentials
    # IMPORTANT: Replace with your actual credentials
    email = "your-email@example.com"
    password = "your-password"
    
    try:
        print("Logging in...")
        user_data = client.login(email, password)
        print(f"[OK] Logged in as: {user_data.get('DisplayName', 'User')}")
        print()
        
        # Get all locations
        print("Fetching locations...")
        locations = client.get_locations()
        print(f"[OK] Found {len(locations)} location(s)")
        print()
        
        for location in locations:
            print(f"Location: {location.name}")
            print(f"  ID: {location.id}")
            print(f"  City: {location.city}, {location.country}")
            print(f"  Zones: {len(location.zones)}")
            print()
            
            # Get detailed system information for this location
            print(f"Fetching system details for {location.name}...")
            system = client.get_location_system(location.id)
            
            # Display zone information
            for zone in system.zones:
                status = "[Online]" if zone.is_alive else "[Offline]"
                override = " (Override Active)" if zone.override_active else ""
                
                print(f"  Zone: {zone.name} {status}{override}")
                print(f"    Current: {zone.temperature}°C")
                print(f"    Target:  {zone.target_temperature}°C")
                print(f"    Range:   {zone.min_temperature}°C - {zone.max_temperature}°C")
                
                if zone.has_alerts:
                    print(f"    [WARNING] Has alerts!")
                
                print()
            
            # Example: Set temperature for a specific zone
            if system.zones:
                example_zone = system.zones[0]
                new_temp = 21.0
                
                print(f"Example: Setting {example_zone.name} to {new_temp}°C...")
                try:
                    client.set_zone_temperature(
                        zone_id=example_zone.id,
                        temperature=new_temp,
                        permanent=True,
                    )
                    print(f"[OK] Temperature set successfully!")
                except APIError as e:
                    print(f"[ERROR] Failed to set temperature: {e}")
                print()
        
        # Get account information
        print("Fetching account information...")
        account = client.get_account_info()
        print(f"[OK] Account: {account.first_name} {account.last_name}")
        print(f"  Email: {account.username}")
        print(f"  Location: {account.city}, {account.country_name}")
        print()
        
    except AuthenticationError as e:
        print(f"[ERROR] Authentication failed: {e}")
        print("Please check your email and password.")
    except ZoneNotFoundError as e:
        print(f"[ERROR] Zone not found: {e}")
    except APIError as e:
        print(f"[ERROR] API error: {e}")
    except Exception as e:
        print(f"[ERROR] Unexpected error: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()

File: rust/examples/basic_usage.rs
//! Basic usage example for the MyTotalConnectComfort API client.

use evohome_rs::{Client, Error};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Create a client instance
    let mut client = Client::new();

    // Login with your credentials
    // IMPORTANT: Replace with your actual credentials
    let email = "your-email@example.com";
    let password = "your-password";

    println!("Logging in...");
    match client.login(email, password).await {
        Ok((user_data, _cookies)) => {
            println!("[OK] Logged in as: {}", user_data.display_name.as_deref().unwrap_or("Unknown"));
        }
        Err(Error::Authentication(msg)) => {
            eprintln!("[ERROR] Authentication failed: {}", msg);
            eprintln!("Please check your email and password.");
            return Ok(());
        }
        Err(e) => return Err(e.into()),
    }

    // Get all locations
    println!("\nFetching locations...");
    let locations = client.get_locations().await?;
    println!("[OK] Found {} location(s)\n", locations.len());

    for location in &locations {
        println!("Location: {}", location.name.as_deref().unwrap_or("Unknown"));
        println!("  ID: {}", location.id);
        if let Some(city) = &location.city {
            if let Some(country) = &location.country {
                println!("  Address: {}, {}", city, country);
            }
        }
        println!("{}: {} zones", location.name.as_deref().unwrap_or("Unknown"), location.zones.len());

        // Get detailed system information for this location
        println!("Fetching system details for {}...", location.name.as_deref().unwrap_or("Unknown"));
        let system = client.get_location_system(&location.id).await?;

        // Display zone information
        for zone in &system.zones {
            let status = if zone.is_alive {
                "[Online]"
            } else {
                "[Offline]"
            };
            let override_info = if zone.override_active {
                " (Override Active)"
            } else {
                ""
            };

            println!("  Zone: {}{}{}", zone.name.as_deref().unwrap_or("Unknown"), status, override_info);
            println!("    Current: {}°C", zone.temperature);
            println!("    Target:  {}°C", zone.target_heat_temperature);
            println!(
                "    Range:   {}°C - {}°C",
                zone.min_heat_setpoint, zone.max_heat_setpoint
            );

            if zone.has_alerts {
                println!("    [WARNING] Has alerts!");
            }
            println!();
        }

        // Example: Set temperature for a specific zone
        if let Some(example_zone) = system.zones.first() {
            let new_temp = 21.0;

            println!(
                "Example: Setting {} to {}°C...",
                example_zone.name.as_deref().unwrap_or("Unknown"), new_temp
            );
            match client
                .set_zone_temperature(&example_zone.id, new_temp, true, 0, 0, false)
                .await
            {
                Ok(_) => println!("[OK] Temperature set successfully!"),
                Err(e) => eprintln!("[ERROR] Failed to set temperature: {}", e),
            }
            println!();
        }
    }

    // Get account information
    println!("Fetching account information...");
    let account = client.get_account_info().await?;
    println!(
        "[OK] Account: {} {}",
        account.first_name.as_deref().unwrap_or(""), account.last_name.as_deref().unwrap_or("")
    );
    println!("  Email: {}", account.username);
    if let (Some(city), Some(country)) = (&account.city, &account.country_name) {
        println!("  Location: {}, {}", city, country);
    }

    Ok(())
}
